File: LAB3-main/funcoes.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include "header.hpp"

using namespace std;

int R = 256;

int charAT(string a,int d){
    if(d < a.size())
        return a.at(d);
    else
        return -1;
}

void MSD_sort(vector<string> &a, vector<string> &aux, int lo, int hi, int d){
    if(hi<=lo)
        return;
    int count[R+2] = {0};
    for(int i = lo;i <= hi; i++)
        count[charAT(a[i],d) + 2]++;
    for(int r = 0; r < R+1;r++)
        count[r+1] += count[r];
    for(int i = lo;i <= hi; i++){
        aux[count[charAT(a[i],d) + 1]++] = a[i];
    }
    for(int i = lo;i <= hi; i++){
        a[i] = aux[i-lo];
    }
    for(int r = 0; r < R;r++)
        MSD_sort(a,aux,lo+count[r],lo+count[r+1] - 1,d+1);
}

vector<string> MSD_start(vector<string> &a){
    int n = a.size();
    vector<string> aux(n);
    MSD_sort(a,aux,0,n-1,0);
    return a;
}

void operation(string ent, string out){
	vector<string> words;
	string str;
	int tam = 0,count = 1;
	ifstream arq_ent;
	ofstream arq_sai;

	arq_ent.open(ent);
	if(arq_ent.is_open()){
		while(arq_ent >> str){
			if(str.length()>=4){
				words.push_back(str);
				tam++;
			}
		}
	}
	else{
		cout << "Nao foi possivel abrir o arquivo" << endl;
	}
	arq_ent.close();
	words = MSD_start(words);
	arq_sai.open(out);
	for(int i = 0; i < tam-1;i++){
		if(words[i] == words[i+1])
			count++;
		else{
			arq_sai << words[i] << "\t" << count << endl;
			count = 1;
		}
	}
	arq_sai.close();
}

File: LAB3-main/funcoes.py
r = 256
m = 15

def charAt(a,d):
    if d < len(a):
        return ord(a[d])
    else:
        return -1

def MSD_start(a):
    n = len(a)
    aux = [0 for i in range(n)]
    MSD_sort(a,aux,0,n-1,0)
    return a

def MSD_sort(a,aux,lo,hi,d):
    if hi <= lo:
        return
    count = [0 for i in range(r+2)]

    for i in range(lo,hi+1):
        count[charAt(a[i],d)+2] += 1
    for j in range(0,r+1):
        count[j+1] += count[j]
    for i in range(lo,hi+1):
        aux[count[charAt(a[i],d)+1]] = a[i]
        count[charAt(a[i],d)+1] +=1
    for i in range(lo,hi+1):
        a[i] = aux[i-lo]
    for j in range(0,r):
        MSD_sort(a,aux,lo + count[j], lo + count[j+1] - 1,d+1)

def operation(nome_ent, nome_sai):
    arq_ent = open(nome_ent,'r')
    a = arq_ent.read().split()
    arq_ent.close()
    i = 0
    while i < len(a):
        if len(a[i]) < 4:
            del a[i]
            i -= 1
        i = i + 1
    k = MSD_start(a)
    count = 1
    arq_sai = open(nome_sai,'w')
    for j in range (len(k)-1):
        if k[j] == k[j+1]:
            count += 1
        else:
            arq_sai.write('{}\t{}\n'.format(k[j],count))
            count = 1
    arq_sai.close()

